README.md
#  Emote Resizer

[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/aosterwyk/emote-resizer?sort=semver)](https://github.com/aosterwyk/emote-resizer/releases) [![GitHub last commit](https://img.shields.io/github/last-commit/aosterwyk/emote-resizer)](https://github.com/aosterwyk/emote-resizer/commits/master) [![GitHub last commit (branch)](https://img.shields.io/github/last-commit/aosterwyk/emote-resizer/develop?label=last%20commit%20%28dev%29)](https://github.com/aosterwyk/emote-resizer/commits/dev) [![Discord](https://img.shields.io/discord/90687557523771392?color=000000&label=%20&logo=discord)](https://discord.gg/QNppY7T) [![Twitch Status](https://img.shields.io/twitch/status/varixx?label=%20&logo=twitch)](https://twitch.tv/VariXx) 

Emote resizer is a tool that resizes .png files to Twitch emote sizes. The name isn't very creative. 

![preview](https://acceptdefaults.com/emote-resizer/emote-resizer-preview.gif)

## Installation

Download and run the installer linked in the [latest release](https://github.com/aosterwyk/emote-resizer/releases) or build using the instructions below.

### Building  
- Clone the repository 
- Install electron-builder (global recommended)
``npm install -g electron-builder``
- Install dependencies 
``npm install`` 
- Run electron-builder from the project directory 
``electron-builder``

## Usage

- Check emote and/or badge checkboxes
- (Optional) Select resize mode
- Drag a .png file to the box labled "Drag and drop your image(s) here"
- Just wait a moment 😎
- You will see 3-6 new files with their resolutons (ex: 28x28, 56x56, etc.) added to their names. 

## Support

[Discord server](https://discord.gg/QNppY7T) or DM `VariXx#8317`

## License
[Apache License 2.0](https://choosealicense.com/licenses/apache-2.0/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
_No unreleased changes_

## [0.7.2] - 2021-11-18
### Changed
- Changed wiki link on help button
- Changed icon

## [0.7.1] - 2021-11-18
### Changed
- Changed year in license 

### Security 
- Updated packages with non-breaking updates

## [0.7.0] - 2020-08-28
### Added
- Added [Apache License 2.0](https://choosealicense.com/licenses/apache-2.0/) license info

### Changed
- Changed Bootstrap theme to [Sketchy](https://bootswatch.com/sketchy/)
- Changed background color

### Fixed
- Fixed changelog formatting

## [0.6.2] - 2020-08-25
### Changed
- Changed height resize back to fixed size

## [0.6.1] - 2020-08-25
### Changed
- Changed height resize to auto size

## [0.6.0] - 2020-08-25
### Added
- Added resize mode options
- Added output size options for emotes (28px, 56px, 112px) and badges (18px, 36px, 72px)
- Added an image icon to drag area
- Added status messages

### Changed
- Changed image resizing library to Jimp

## [0.5.0] - 2020-08-23
### Fixed
- Fixed help button
- Fixed discord and help button positions

### Removed
- Removed menu bar from window

## [0.4.0] - 2020-08-23
### Changed
- Incremented version to test autoupdate

## [0.3.0] - 2020-08-23
### Changed
- Changed product name (installer) and window title

## [0.2.0] - 2020-08-23
### Added
- Added UI 

## [0.1.0] - 2020-08-23
### Added
- Added function for resizing images

### Initial Release

[0.7.2]: https://github.com/aosterwyk/emote-resizer/tree/v0.7.2
[0.7.1]: https://github.com/aosterwyk/emote-resizer/tree/v0.7.1
[0.7.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.7.0
[0.6.2]: https://github.com/aosterwyk/emote-resizer/tree/v0.6.2
[0.6.1]: https://github.com/aosterwyk/emote-resizer/tree/v0.6.1
[0.6.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.6.0
[0.5.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.5.0
[0.4.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.4.0
[0.3.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.3.0
[0.2.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.2.0
[0.1.0]: https://github.com/aosterwyk/emote-resizer/tree/v0.1.0
[Unreleased]: https://github.com/aosterwyk/emote-resizer/compare/master...develop


main.js
const fs = require('fs');
const Jimp = require('jimp');
const { autoUpdater } = require('electron-updater');
const { app, BrowserWindow, ipcMain, ipcRenderer } = require('electron');
const emoteSizes = [112, 56, 28];
var win;

// const testImg = 'test.png'; // test file 

function sendStatusMsg(newMsgType, newMsg) {
    let sendMsg = {
        msgType: newMsgType,
        message: newMsg
    }
    win.webContents.send('statusMessage', sendMsg);
    console.log(newMsg);
}

async function resize(imgName, newSizeWidth, newSizeHeight, newResizeMode) {
    if(imgName.includes('.png')) {
        // console.log(`Reading image ${imgName}`);
        sendStatusMsg(`info`,`Reading image ${imgName}`);
        const sourceImg = await Jimp.read(`${imgName}`);
        // console.log(`Resizing image ${imgName}`);
        sendStatusMsg(`info`,`Resizing image ${imgName}`);
        let resizeMode = Jimp.RESIZE_BILINEAR; 
        if(newResizeMode !== `default`) {
            if(newResizeMode == `nearestNeighbor`) {
                resizeMode = Jimp.RESIZE_NEAREST_NEIGHBOR;
            }
            else if(newResizeMode == `bilinear`) {
                resizeMode = Jimp.RESIZE_BILINEAR;
            }
            else if(newResizeMode == `bicubic`) {
                resizeMode = Jimp.RESIZE_BICUBIC;
            }
            else if(newResizeMode == `hermite`) {
                resizeMode = Jimp.RESIZE_HERMITE;
            }
            else if(newResizeMode == `bezier`) {
                resizeMode = Jimp.RESIZE_BEZIER;
            }
        }
        sourceImg.resize(newSizeWidth, newSizeHeight, resizeMode);
        // console.log(`${imgName} resized to ${newSizeWidth} x ${newSizeHeight} (${resizeMode})`);
        sendStatusMsg(`success`, `${imgName} resized to ${newSizeWidth} x ${newSizeHeight} (${resizeMode})`);
        let splitImgName = imgName.split(".png");
        let newFilename = `${splitImgName[0]}-${newSizeHeight}x${newSizeWidth}.png`;
        try {
            await sourceImg.writeAsync(newFilename);
        }
        catch(error) {
            sendStatusMsg(`error`, `Error saving image ${newFilename}: ${error}`);
        }
        // console.log(`Resized image saved to ${newFilename}`);
        sendStatusMsg(`success`, `Resized image saved to ${newFilename}`);
    }
    else { sendStatusMsg(`error`, 'Filename does not end in .png. Skipping...'); }
}

function createWindow() { 
    win = new BrowserWindow({
        width: 1000,
        height: 800,
        webPreferences: {
            nodeIntegration: true
        }
    });

    win.loadFile('index.html');
    win.menuBarVisible = false;
    // win.webContents.openDevTools();
}

app.whenReady().then(() => {
    createWindow();
    app.on('activate', () => {
        if(BrowserWindow.getAllWindows().length === 0) {
            createWindow();
        }
    });
});

app.on('window-all-closed', () => {
    if(process.platform !== 'darwin') {
        app.quit();
    }
});

app.on('ready', () => {
    autoUpdater.checkForUpdatesAndNotify();
});

ipcMain.handle('draggedItem', async (event, args) => {
    for(let i = 0; i < args.resolutions.length; i++) {
        await resize(args.filePath, args.resolutions[i], args.resolutions[i], args.resizeMode);
    }
});


render.js
const { ipcRenderer, shell } = require('electron');

function updateStatus(newMsgType, newMsg) {
    let oldStatus = document.getElementById(`statusMsg`);
    if(newMsgType === `error`) {
        msgColor = `text-danger`;
    }
    else if(newMsgType === `success`) {
        msgColor = `text-success`;
    }
    else {
        msgColor = `text-muted`;
    }
    oldStatus.innerHTML = `<small class='${msgColor} text-xl-left'>${newMsg}</small><br>${oldStatus.innerHTML}`;
    console.log(newMsg);
}

ipcRenderer.on(`statusMessage`, (event, msg) => {
    updateStatus(msg.msgType, msg.message);
});

document.addEventListener('drop', (event) => { 
	event.preventDefault(); 
	event.stopPropagation(); 
    for(const file of event.dataTransfer.files) {
        updateStatus(`info`, `Loaded ${file.path}`);
        
        let outputSizes = [];
        if(document.getElementById(`emotes`).checked) { 
            outputSizes.push(28, 56, 112);
        }
        if(document.getElementById(`badges`).checked) {
            outputSizes.push(18, 36, 72);
        }

        let resizeMode = ``;
        let resizeModes = document.getElementsByName(`resizeModeChoice`);
        for(let i=0; i < resizeModes.length; i++) {
            if(resizeModes[i].checked) { 
                resizeMode = resizeModes[i].value;
            }
        }
        updateStatus(`info`, `Using resize mode: ${resizeMode}`);

        let resizeJob = {
            filePath: file.path,
            resizeMode: resizeMode,
            resolutions: outputSizes
        }
        ipcRenderer.invoke('draggedItem', resizeJob);
    }
}); 

document.addEventListener('dragover', (e) => { 
	e.preventDefault(); 
	e.stopPropagation(); 
}); 

function externalLink(destination) {
    if(destination == `discord`) {
        shell.openExternal(`https://discord.gg/QNppY7T`);
    }
    if(destination == `help`) {
        shell.openExternal(`https://github.com/aosterwyk/emote-resizer/wiki`); 
    }
}
